crates/astria-sequencer/src/accounts/storage/values.rs

```diff
@@ -0,0 +1,128 @@
+use std::fmt::{
+    self,
+    Display,
+    Formatter,
+};
+
+use astria_eyre::eyre::bail;
+use borsh::{
+    BorshDeserialize,
+    BorshSerialize,
+};
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub(crate) struct Value(ValueImpl);
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+enum ValueImpl {
+    Balance(Balance),
+    Nonce(Nonce),
+    Fee(Fee),
+}
+
+impl Display for Value {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            ValueImpl::Balance(balance) => write!(f, "balance {}", balance.0),
+            ValueImpl::Nonce(nonce) => write!(f, "nonce {}", nonce.0),
+            ValueImpl::Fee(fee) => write!(f, "fee {}", fee.0),
+        }
+    }
+}
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub(in crate::accounts) struct Balance(u128);
+
+impl From<u128> for Balance {
+    fn from(balance: u128) -> Self {
+        Balance(balance)
+    }
+}
+
+impl From<Balance> for u128 {
+    fn from(balance: Balance) -> Self {
+        balance.0
+    }
+}
+
+impl<'a> From<Balance> for crate::storage::StoredValue<'a> {
+    fn from(balance: Balance) -> Self {
+        crate::storage::StoredValue::Accounts(Value(ValueImpl::Balance(balance)))
+    }
+}
+
+impl<'a> TryFrom<crate::storage::StoredValue<'a>> for Balance {
+    type Error = astria_eyre::eyre::Error;
+
+    fn try_from(value: crate::storage::StoredValue<'a>) -> Result<Self, Self::Error> {
+        let crate::storage::StoredValue::Accounts(Value(ValueImpl::Balance(balance))) = value
+        else {
+            bail!("accounts stored value type mismatch: expected balance, found {value}");
+        };
+        Ok(balance)
+    }
+}
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub(in crate::accounts) struct Nonce(u32);
+
+impl From<u32> for Nonce {
```

Comment:
And this should also have a domain type.

Reply:
Ticketed at #1605.